### PR TITLE
Add AddAutoMapper overload with IServiceProvider

### PR DESCRIPTION
**Pull request by Tasteful**
Add overloads that exposing the `IServiceProvider` during configuration of the AutoMapper.

See
https://github.com/AutoMapper/AutoMapper.Collection.EFCore/issues/2#issuecomment-412267189

### CODE REVIEW

**Comment by ChristopherHaws** [comment]
I like this. I also feel like `IServiceProvider ApplicationServices { get; }` should be a property on `IMapperConfigurationExpression`. The default implementation could be an `ActivatorServiceProvider` for static usages of AutoMapper and `AddAutoMapper` could set it to be the `IServiceProvider`.

This would allow for the following use (this is somewhat sudo code):
```csharp
/* User's Application */
void Main()
{
    var services = new ServiceCollection();

    services.AddAutoMapper(automapper =>
    {
        automapper.UseCollectionMappers();
        automapper.UseEntityFrameworkCoreModel<BlogContext>();
    });
}

public class BlogContext : DbContext
{
}

/* AutoMapper */

public interface IMapperConfigurationExpression
{
    IServiceProvider ApplicationServiceProvider { get; }
}

public class MapperConfigurationExpression : IMapperConfigurationExpression
{
    public MapperConfigurationExpression(IServiceProvider applicationServiceProvider) =>
        this.ApplicationServiceProvider = applicationServiceProvider;

    public IServiceProvider ApplicationServiceProvider { get; private set; }
}

/* AutoMapper.Extensions.Microsoft.Extensions */

public static class AutoMapperServiceCollectionExtensions
{
    public static IServiceCollection AddAutoMapper(this IServiceCollection services, Action<IMapperConfigurationExpression> optionsAction = null)
    {
        services.AddSingleton<IMapperConfigurationExpression>(serviceProvider =>
        {
            var options = new MapperConfigurationExpression(serviceProvider);
            optionsAction?.Invoke(options);
            
            return options;
        });
        
        return services;
    }
}

/* AutoMapper.Collection */

public static class MapperConfigurationExpressionCollectionExtensions
{
    public static void UseCollectionMappers(this MapperConfigurationExpression options) { }
    public static void SetGeneratePropertyMaps(this MapperConfigurationExpression options, IGeneratePropertyMaps maps) { }
}

public interface IGeneratePropertyMaps { }

/* AutoMapper.EntityFrameworkCore */
public static class MapperConfigurationExpressionEntityFrameworkCoreExtensions
{
    public static void UseEntityFrameworkCoreModel<TContext>(this MapperConfigurationExpression options)
        where TContext : DbContext
    {
        using (var scope = options.ApplicationServiceProvider.CreateScope())
        {
            var db = scope.ServiceProvider.GetRequiredService<TContext>();
            options.SetGeneratePropertyMaps(new GenerateEntityFrameworkCorePrimaryKeyPropertyMaps<TContext>(db.Model));
        }
    }
}

public class GenerateEntityFrameworkCorePrimaryKeyPropertyMaps<TContext> : IGeneratePropertyMaps
    where TContext : DbContext
{
    public GenerateEntityFrameworkCorePrimaryKeyPropertyMaps(IModel model) { }
}
```

**Comment by Tasteful** [comment]
I was first trying to rebuild so the `Action<IMapperConfigurationExpression>` was an `Action<AutoMapperConfigBuilder>` that was inheriting both the `IMapperConfigurationExpression` and `IServiceProvider`.

**Comment by ChristopherHaws** [comment]
`IServiceProvider` is part of the `System` namespace, so no dependencies would need to be added. As a side note, `IMapperConfigurationExpression`'s implementation is `MapperConfigurationExpression` which also implements `IConfiguration`. `IConfiguration` actually has a service factory `Func<Type, object> ServiceCtor` which by default uses `Activator`. Could this somehow be used? If `cfg.ConstructServicesUsing(serviceProvider.GetService);` were added before [this line](https://github.com/Tasteful/AutoMapper.Extensions.Microsoft.DependencyInjection/blob/AddAutoMapper-with-serviceProvider/src/AutoMapper.Extensions.Microsoft.DependencyInjection/ServiceCollectionExtensions.cs#L83) then it probably could be used during configuration.

**Comment by Tasteful** [comment]
The `IServiceProvider` is part of system namespace but it is hard to say that you always need to have an implementation for that interface to be able to use AM. That will be an extra dependency (indirectly) that AM not is needing.

The method `ServiceCtor` was used to resolve the `IServiceProvider` (in the AM.Collection.Efcore that you removed in you PR). The problem with that method is that the `IMapper` is created with an overload that will pass in an scoped `IServiceProvider`, assigning an method to the `ServiceCtor` will make it to use a Singleton instance that will limit the usecase.

File: src/AutoMapper.Extensions.Microsoft.DependencyInjection/ServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -103,14 +103,12 @@ void ConfigAction(IMapperConfigurationExpression cfg)
                 services.AddTransient(type.AsType());
             }
 
-            services.AddSingleton(config);
+            services.AddSingleton<IConfigurationProvider>(sp => new MapperConfiguration(cfg => ConfigAction(sp, cfg)));
```
Comment (jbogard) [inline src/AutoMapper.Extensions.Microsoft.DependencyInjection/ServiceCollectionExtensions.cs]: What's the implication here of moving from an instance to a lambda?

Comment (Tasteful) [reply src/AutoMapper.Extensions.Microsoft.DependencyInjection/ServiceCollectionExtensions.cs]: We are moving the creation of `IConfigurationProvider` from when we call `AddAutoMapper`-extension into when it is used for the first time.

**Comment by Tasteful** [comment]
@jbogard Can you bump version number, the myget created one have same version number as the released one.

**Comment by jbogard** [comment]
Done!

**Pull request merged**